### PR TITLE
Feature/109034 consume person messages from bus

### PR DESCRIPTION
**Pull request by trondlindanger**
[AB#109034](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/109034)

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Domain/AggregateModels/PersonAggregate/Person.cs [hunk 1]
```diff
@@ -12,14 +12,25 @@ public class Person : EntityBase, IAggregateRoot, IModificationAuditable, IHaveG
...
     public const int EmailLengthMax = 128;
 
-    public Person(Guid guid, string firstName, string lastName, string userName, string email, bool superuser)
+#pragma warning disable CS8618
+    protected Person()
+#pragma warning restore CS8618
+    {
+    }
+
+    public Person(Guid guid, string firstName, string lastName, string userName, string email, bool superuser, DateTime? lastUpdated = null)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Domain/AggregateModels/PersonAggregate/Person.cs:21]: DateTime lastUpdated, and remove if at line 29, and set ProCoSys4LastUpdated = lastUpdated

Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Domain/AggregateModels/PersonAggregate/Person.cs:21]: Update PersonTest as this:
image

Update ProjectTest.cs in same manner

File: src/Equinor.ProCoSys.Completion.WebApi/DiModules/ApplicationModule.cs
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/DiModules/ApplicationModule.cs:1]: Remove using Equinor.ProCoSys.Completion.Command.EventHandlers.DomainEvents.PunchItemEvents.IntegrationEvents; at line 10

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        person.FirstName = personEvent.FirstName;
+        person.LastName = personEvent.LastName;
+        person.Email = personEvent.Email;
+        person.UserName = personEvent.UserName;
+        person.Superuser = personEvent.SuperUser;
+        person.ProCoSys4LastUpdated = personEvent.LastUpdated;
+    }
+
+    private static Person CreatePersonEntity(IPersonEventV1 personEvent)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs:75]: remove

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                context.MessageId, personEvent.Guid, personEvent.LastUpdated, person.ProCoSys4LastUpdated);
+            return;
+        }
+
+        MapFromEventToPerson(personEvent, person);
+
+        _currentUserSetter.SetCurrentUserOid(_options.CurrentValue.CompletionApiObjectId);
+        await _unitOfWork.SaveChangesAsync(context.CancellationToken);
+
+        _logger.LogInformation("Person Message Consumed: {MessageId} \n {UserName}",
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs:61]: Also add personEvent.Guid to log statement (do the same in ProjectEventConsumer)

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    public async Task Consume(ConsumeContext<PersonEvent> context)
+    {
+        var personEvent = context.Message;
+
+        if (personEvent.Guid == Guid.Empty)
+        {
+            throw new Exception("Message is missing user Azure oid");
+        }
+
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs:43]: Add code
```
        if (!await _personRepository.ExistsAsync(personEvent.Guid, context.CancellationToken))
        {
            _logger.LogInformation("Person not exists. Message Skipped: {MessageId} \n {UserName} \n {Guid}",
                context.MessageId, personEvent.UserName, personEvent.Guid);
            return;
        }
```

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/PersonEventConsumerTests.cs
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/PersonEventConsumerTests.cs:1]: Add test: Consume_ShouldIgnoreMessage_WhenPersonDoesNotExist()

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/PersonEventConsumerTests.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Threading.Tasks;
+using Equinor.ProCoSys.Common.Misc;
+using Equinor.ProCoSys.Completion.Domain;
+using Equinor.ProCoSys.Completion.Domain.AggregateModels.PersonAggregate;
+using Equinor.ProCoSys.Completion.Domain.AggregateModels.ProjectAggregate;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/PersonEventConsumerTests.cs:6]: remove

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/PersonEventConsumerTests.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+[TestClass]
+public class PersonEventConsumerTests
+{
+    private readonly IPersonRepository _personRepoMock = Substitute.For<IPersonRepository>();
+    private readonly IUnitOfWork _unitOfWorkMock = Substitute.For<IUnitOfWork>();
+    private readonly PersonEventConsumer _personEventConsumer;
+    private readonly IOptionsMonitor<CompletionAuthenticatorOptions> _optionsMock = Substitute.For<IOptionsMonitor<CompletionAuthenticatorOptions>>();
+    private readonly ConsumeContext<PersonEvent> _contextMock = Substitute.For<ConsumeContext<PersonEvent>>();
+    private Project? _projectAddedToRepository;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/PersonEventConsumerTests.cs:26]: remove

File: src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/PersonEventConsumerTests.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    public PersonEventConsumerTests() =>
+        _personEventConsumer = new PersonEventConsumer(Substitute.For<ILogger<PersonEventConsumer>>(), _personRepoMock, 
+            _unitOfWorkMock, Substitute.For<ICurrentUserSetter>(), _optionsMock);
+
+    [TestInitialize]
+    public void Setup()
+    {
+        _optionsMock.CurrentValue.Returns(new CompletionAuthenticatorOptions { CompletionApiObjectId = new Guid() });
+        
+        _personRepoMock
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.Tests/Synchronization/PersonEventConsumerTests.cs:37]: remove

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+public record PersonEvent(string EventType, 
+        Guid Guid, 
+        string FirstName,
+        string LastName,
+        string UserName,
+        string Email,
+        bool SuperUser,
+        DateTime LastUpdated,
+        string? Behavior)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs:93]: remove string? Behavior

**Comment by janid1967** [comment]
wait for feedback

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs (hunk 2 above)
Comment (trondlindanger) [reply src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs:75]: We do not want to persist last updated through new ProCoSys4LastUpdated column in Person table?

Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs:75]: remove private static Person CreatePersonEntity ... not in use

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs (hunk 8 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PersonEventConsumer.cs:93]: As discussed, keep string? Behavior but implement a check if it is "delete". On "delete", log it and return without doing anything

File: src/Equinor.ProCoSys.Completion.Domain/AggregateModels/PersonAggregate/Person.cs (hunk 1 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Domain/AggregateModels/PersonAggregate/Person.cs:21]: As discussed, remove DateTime lastUpdated as param in constructor. Set ProCoSys4LastUpdated = DateTime.Now in constructor

Make property ProCoSys4LastUpdated  private and create a method SetProCoSys4LastUpdated to set it

Perform same change on Project

Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Domain/AggregateModels/PersonAggregate/Person.cs:21]: Create an unit test on SetProCoSys4LastUpdated

Comment (trondlindanger) [reply src/Equinor.ProCoSys.Completion.Domain/AggregateModels/PersonAggregate/Person.cs:21]: Direct unit test of a setter method? Do you mean indirect testing?

Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Domain/AggregateModels/PersonAggregate/Person.cs:21]: We can talk about it

**Review by janid1967 — APPROVED** [review]

**Pull request merged**